mm2src/coins/lightning.rs

```diff
@@ -508,33 +830,122 @@ impl MarketCoinOps for LightningCoin {
         ))
     }
 
-    // Todo: Implement this when implementing swaps for lightning as it's is used mainly for swaps
+    // Todo: Add waiting for confirmations logic for the case of if the channel is closed and the htlc can be claimed on-chain
     fn wait_for_confirmations(
         &self,
-        _tx: &[u8],
+        tx: &[u8],
         _confirmations: u64,
         _requires_nota: bool,
-        _wait_until: u64,
-        _check_every: u64,
+        wait_until: u64,
+        check_every: u64,
     ) -> Box<dyn Future<Item = (), Error = String> + Send> {
-        unimplemented!()
+        let payment_hash = try_f!(payment_hash_from_slice(tx).map_err(|e| e.to_string()));
+        let payment_hex = hex::encode(payment_hash.0);
+
+        let coin = self.clone();
+        let fut = async move {
+            loop {
+                if now_ms() / 1000 > wait_until {
+                    return ERR!(
+                        "Waited too long until {} for payment {} to be received",
+                        wait_until,
+                        payment_hex
+                    );
+                }
+
+                match coin.db.get_payment_from_db(payment_hash).await {
+                    Ok(Some(_)) => {
```

Comment:
`wait_for_confirmations` does 2 jobs when used in swaps currently, it first waits for the transaction to appear on-chain then it waits for the required confirmations, that's why we got an issue like this https://github.com/KomodoPlatform/atomicDEX-API/issues/1473.
But since for lightning there is no concept of confirmations, it's used here to checks for if the payment sent/received/spent by the other side which is equivalent to transactions appearing on-chain, but we need to check for if the payment is sent/received/spent depending on where `wait_for_confirmation` is called in `taker_swap`, `maker_swap`.
That's why I propose to breaking it down to 3 functions `wait_for_payment_sent_confirmations`, `wait_for_payment_received_confirmations`, `wait_for_payment_spent_confirmations` but this should be part of this discussion https://github.com/KomodoPlatform/atomicDEX-API/issues/1501
Currently taker as lightning is only implemented, so this just check if the payment is in db since for the maker the payment won't appear on db unless it's received
https://github.com/KomodoPlatform/atomicDEX-API/blob/60acd0af21dda65dbe8f43ca9a1fa6994b22571d/mm2src/coins/lightning/ln_events.rs#L373 As for the taker this https://github.com/KomodoPlatform/atomicDEX-API/blob/60acd0af21dda65dbe8f43ca9a1fa6994b22571d/mm2src/mm2_main/src/lp_swap/taker_swap.rs#L1453 will pass straight away since the payment is saved to db when sending the taker payment here https://github.com/KomodoPlatform/atomicDEX-API/blob/60acd0af21dda65dbe8f43ca9a1fa6994b22571d/mm2src/coins/lightning.rs#L483
https://github.com/KomodoPlatform/atomicDEX-API/blob/60acd0af21dda65dbe8f43ca9a1fa6994b22571d/mm2src/coins/lightning.rs#L226
Then `wait_for_tx_spend` https://github.com/KomodoPlatform/atomicDEX-API/blob/60acd0af21dda65dbe8f43ca9a1fa6994b22571d/mm2src/mm2_main/src/lp_swap/taker_swap.rs#L1472 will wait until the status changes to `Succeeded` here https://github.com/KomodoPlatform/atomicDEX-API/blob/60acd0af21dda65dbe8f43ca9a1fa6994b22571d/mm2src/coins/lightning/ln_events.rs#L440
I can add a new parameter to `wait_for_confirmation` to pass what we need to check for with every call of the function, but the name `wait_for_confirmation` is not very good actually for the functions multiple purposes now.

So to conclude this long explanation, this will be better improved in upcoming PRs, that's why I left the `Todo` inside `wait_for_confirmation` implementation for `lightning ` for now.

cc @artemii235 @ozkanonur since this comment is related to this discussion https://github.com/KomodoPlatform/atomicDEX-API/issues/1501

Reply:
I see, thanks for such a comprehensive answer!